### ISSUE TITLE
Recursive reference in custom serializer

### ISSUE THREAD

**Issue by gjeraj** [eclipse-ee4j/yasson#187]
I'm trying to implement a custom abstract class serializer which could serialize POJO in the following way: `{"com.test.TestClass":{"a":"1","b":"2"}}`

For that I've written a custom serializer which is identical to the one listed in JSON-B example (http://json-b.net/docs/user-guide.html#serializers-deserializers):

```java
public class CustomSerializer implements JsonbSerializer<GenericClass> {
	@Override
	public void serialize(GenericClass instance, JsonGenerator generator, SerializationContext context) {
		generator.writeStartObject();

		if (instance != null)
			context.serialize(instance.getClass().getName(), instance, generator);
		else
			context.serialize(null, generator);

		generator.writeEnd();
	}
}
```

When i try to run it (`jsonb.toJson(instance, GenericClass.class);`) i get `javax.json.bind.JsonbException: Recursive reference has been found in class class com.test.TestClass.` 

Funny thing is that this used to work in yasson:1.0.1. The problem in 1.0.1 is when you call `.toJson` a second time on the same instance of `Jsonb` it would ignore custom serializer. Is there any way to get around this problem?

I've written a small test project to help: https://github.com/gjeraj/test-yasson

Thanks

**Comment by aguibert** [comment]
@bravehorsie this seems to be an issue caused by XYZ

Also, @gjeraj has pointed out we have an invalid code example in the [JSON-B doc](http://json-b.net/docs/user-guide.html#serializers-deserializers):
```java
public class AnimalSerializer implements JsonbSerializer<Animal> {
    public void serialize(Animal animal, JsonGenerator jsonGenerator, SerializationContext serializationContext) {
        if (animal != null) {
            serializationContext.serialize(animal.getClass().getName(), animal, jsonGenerator);
        } else {
            serializationContext.serialize(null, jsonGenerator);
        }
    }
}
```

With the current code we get:
>  javax.json.stream.JsonGenerationException: Illegal method during JSON generation, not valid in current context IN_NONE

If we add a `generator.writeStartObject()` as shown in the OP, we get the following error:
```
javax.json.bind.JsonbException: Recursive reference has been found in class class atest.AGGMain$TestClass.
	at org.eclipse.yasson.internal.serializer.UserSerializerSerializer.serialize(UserSerializerSerializer.java:57)
	at org.eclipse.yasson.internal.Marshaller.serializeRoot(Marshaller.java:118)
	at org.eclipse.yasson.internal.Marshaller.serialize(Marshaller.java:93)
	at atest.AGGMain$CustomSerializer.serialize(AGGMain.java:86)
	at atest.AGGMain$CustomSerializer.serialize(AGGMain.java:1)
	at org.eclipse.yasson.internal.serializer.UserSerializerSerializer.serialize(UserSerializerSerializer.java:55)
	at org.eclipse.yasson.internal.Marshaller.serializeRoot(Marshaller.java:118)
	at org.eclipse.yasson.internal.Marshaller.marshall(Marshaller.java:74)
	at org.eclipse.yasson.internal.JsonBinding.toJson(JsonBinding.java:106)
```

---

It seems this is expected behavior, introduced by @bravehorsie in this commit: https://github.com/eclipse-ee4j/yasson/commit/5e04a8d77cb18f7a8b59b263276b2023d245454f.  
Perhaps Roman can comment on if this should be allowed or not?

**Comment by bravehorsie** [comment]
@aguibert The commit you have referenced is old and has been overwritten by other commits. Yasson checks for StackOverflowError before it happens in cases such are bidirectional references.

@gjeraj In the test project you referenced there is a `CustomSerializer`, which inside its body calls context.serialize(..). Now if JsonbException with "Recursive .. " text will not be thown what will happen are recursive calls of your CustomSerializer. To Illustrate:

- Yasson runtime serializes root:
  - Found `CustomSerializer` for GenericClass type, calling it
    - `CustomSerializer` passes to Yasson runtime the `GenericClass` object in `context.serialize(..)`.
      - Yasson runtime found a `CustomSerializer` for `GenericClass` object, calling it
        - `CustomSerializer` will continue to pass the object to Yasson runtime causing `StackOverflowError`, if `JsonGenerationException` would not hit first, because you cant create Json object inside other Json object without a json field declared first.

The error is a bit confusing for anyone experimenting custom serializers. Any ideas to make it more selfdescribing?

**Comment by bravehorsie** [comment]
Also related to #164 where custom serializers / adapters were made triggering also for subtypes of its declaration.

**Comment by bravehorsie** [comment]
If there are any new objections I am about to close the issue.

**Comment by rady66** [comment]
It is reproducible on 1.0.2 but not on 1.0.3
`Recursive reference has been found in class class java.util.Date.

<init>:62, JsonbException (javax.json.bind)
serialize:65, AdaptedObjectSerializer (org.eclipse.yasson.internal.serializer)
serializerCaptor:95, AbstractContainerSerializer (org.eclipse.yasson.internal.serializer)
marshallProperty:92, ObjectSerializer (org.eclipse.yasson.internal.serializer)
serializeInternal:61, ObjectSerializer (org.eclipse.yasson.internal.serializer)
serialize:63, AbstractContainerSerializer (org.eclipse.yasson.internal.serializer)
serializeRoot:118, Marshaller (org.eclipse.yasson.internal)
marshall:74, Marshaller (org.eclipse.yasson.internal)
toJson:98, JsonBinding (org.eclipse.yasson.internal)`

**Comment by AlexanderHofmeister** [comment]
Is there any workaround?  

I am trying to hide my entity references behind IDs e.g.

```
public class A {
private B b; 
private C;c
}
```

should lead to something like this: 

```
{
"b": "5",
"c": "2"
}
```

where as "5" and "2" are the referenced IDs 

I implemented something like the code above: 


```
public class MyAdapter implements JsonbSerializer<A> {

    @Override
    public void serialize(A obj, JsonGenerator generator, SerializationContext ctx) {
        generator.writeStartObject("a");
        generator.write("b", obj.getB().getId().toString());
        generator.write("c", obj.getC().getId().toString());
        ctx.serialize(obj, generator);
    }
}

```
I get 

> javax.json.stream.JsonGenerationException: Illegal method during JSON generation, not valid in current context IN_NONE
> at org.glassfish.json.JsonGeneratorImpl.writeStartObject(JsonGeneratorImpl.java:161)
> at [internal classes]

Am I missing something?

**Comment by bravehorsie** [comment]
Are you serializing a root object? If sou you can't call `generator.writeStartObject("a");`, because object properties must be enclosed inside json object curly bracers.

**Comment by AlexanderHofmeister** [comment]
Even without `generator.writeStartObject("a");` I get the same error. :/

**Comment by funcbro** [comment]
The origin of the error seems to be the management of the objects in JsonbContext.

```
    /**
     * Used to avoid StackOverflowError, when adapted / serialized object
     * contains contains instance of its type inside it or when object has recursive reference.
     */
    private Set<Object> currentlyProcessedObjects;
...
    public boolean addProcessedObject(Object object) {
        return this.currentlyProcessedObjects.add(object);
    }
```

Depending on the Hashcode of the processed objects this results in collisions, causing a recursive reference exception in AdaptedObjectSerializer:

```
    @Override
    @SuppressWarnings("unchecked")
    public void serialize(T obj, JsonGenerator generator, SerializationContext ctx) {
        JsonbContext jsonbContext = ((Marshaller) ctx).getJsonbContext();
        try {
            if (jsonbContext.addProcessedObject(obj)) {
                final JsonbAdapter<T, A> adapter = (JsonbAdapter<T, A>) adapterInfo.getAdapter();
                A adapted = adapter.adaptToJson(obj);
                final JsonbSerializer<A> serializer = resolveSerializer((Marshaller) ctx, adapted);
                serializer.serialize(adapted, generator, ctx);
            } else {
                throw new JsonbException(Messages.getMessage(MessageKeys.RECURSIVE_REFERENCE, obj.getClass()));
            }
        } catch (Exception e) {
            throw new JsonbException(Messages.getMessage(MessageKeys.ADAPTER_EXCEPTION, adapterInfo.getBindingType(), adapterInfo.getToType(), adapterInfo.getAdapter().getClass()), e);
        } finally {
            jsonbContext.removeProcessedObject(obj);
        }
    }

```

As a workaround you can try to adapt your hashcode generation for all of your entities.
A solution within Yasson should at least also distinguish the types.

**Comment by masini** [comment]
I have the the same problem with a  JsonbSerializer<List<MyPojo>>, when the object is inside an ArrayList or a wrapper returned by the ORM these are detected as "RECURSIVE_REFERENCE".

Are polymorhpics collections planned in the Jsonb specs ?

**Comment by johnrodey** [comment]
I too have experienced this problem.  Previously (running 2.27) this code would print the root document, however oddly it would only work once if I reused the Jsonb object.  Recreating those objects each time is expensive so I updated to the latest (2.30) and now the same code results in this 'Recursive reference has been found' issue.  So is this supposed to work?  Or if there is another means to serialize the object with the root level can someone provide me a sample?

Below is some trimmed down sample code:

public class CustomSerializer implements JsonbSerializer<SAMPLE> {
	public CustomSerializer() {}
	@Override
	public void serialize(final SAMPLE p, final JsonGenerator g, final SerializationContext c) {
		g.writeStartObject();
		c.serialize(p.getClass().getSimpleName(), p, g);
		g.writeEnd();
	}
}

My object to be serialized:
@JsonbTypeSerializer(CustomSerializer.class)
public class SAMPLE {
	private String id;
	public String getId() {
		return id;
	}
	public void setId(String id) {
		this.id = id;
	}
}

I'm looking to get the output:
{
    "SAMPLE": {
	"id":1
    }
}
